tests/phpunit/integration/Modules/Analytics_4Test.php

```diff
@@ -20,45 +20,142 @@
+use Google\Site_Kit_Dependencies\GuzzleHttp\Stream\Stream;
+use \ReflectionClass;
 
 /**
  * @group Modules
  */
 class Analytics_4Test extends TestCase {
+
 	use Module_With_Scopes_ContractTests;
 	use Module_With_Settings_ContractTests;
 	use Module_With_Owner_ContractTests;
 
+	/**
+	 * Context object.
+	 *
+	 * @var Context
+	 */
+	private $context;
+
+	/**
+	 * Analytics 4 object.
+	 *
+	 * @var Analytics_4
+	 */
+	private $analytics;
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->context   = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$this->analytics = new Analytics_4( $this->context );
+	}
+
 	public function test_register() {
-		$analytics = new Analytics_4( new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE ) );
 		remove_all_filters( 'googlesitekit_auth_scopes' );
 
-		$analytics->register();
+		$this->analytics->register();
 
 		// Adding required scopes.
 		$this->assertEquals(
-			$analytics->get_scopes(),
+			$this->analytics->get_scopes(),
 			apply_filters( 'googlesitekit_auth_scopes', array() )
 		);
 	}
 
-	public function test_get_scopes() {
-		$analytics = new Analytics_4( new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE ) );
+	public function test_handle_provisioning_callback() {
+		$property_id      = '1001';
+		$webdatastream_id = '2001';
+		$measurement_id   = '1A2BCD345E';
+
+		$http_client = new FakeHttpClient();
+		$http_client->set_request_handler(
+			function( Request $request ) use ( $property_id, $webdatastream_id, $measurement_id ) {
+				$url = parse_url( $request->getUrl() );
+				if ( 'analyticsadmin.googleapis.com' !== $url['host'] ) {
+					return new Response( 200 );
+				}
+
+				switch ( $url['path'] ) {
+					case '/v1alpha/properties':
+						return new Response(
+							200,
+							array(),
+							Stream::factory(
+								json_encode(
+									array(
+										'name' => "properties/{$property_id}",
+									)
+								)
+							)
+						);
+					case "/v1alpha/properties/{$property_id}/webDataStreams":
+						return new Response(
+							200,
+							array(),
+							Stream::factory(
+								json_encode(
+									array(
+										'name'          => "properties/{$property_id}/webDataStreams/{$webdatastream_id}",
+										'measurementId' => $measurement_id,
+									)
+								)
+							)
+						);
+					default:
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$this->analytics->get_client()->setHttpClient( $http_client );
+
+		$options   = new Options( $this->context );
+		$analytics = new Analytics_4( $this->context, $options );
+
+		$options->set(
+			Settings::OPTION,
+			array(
+				'accountID'       => '12345678',
+				'propertyID'      => '',
+				'webDataStreamID' => '',
+				'measurementID'   => '',
+			)
+		);
+
+		$handle_provisioning_callback = ( new ReflectionClass( $analytics ) )->getMethod( 'handle_provisioning_callback' );
+		$handle_provisioning_callback->setAccessible( true );
+		$handle_provisioning_callback->invoke( $this->analytics, '12345678' );
```

Comment:
Let's call this via `do_action( 'googlesitekit_analytics_handle_provisioning_callback' ... )` instead. This will require us to register the `Analytics_4` instance here as well.

Reply:
Updated.